### PR TITLE
feat(echo-db): client-side query coalescing to eliminate duplicate QueryContext instances

### PR DESCRIPTION
**Pull request by dmaretskyi**
## Summary

- Introduces `QueryContextCoalescer` that deduplicates identical queries at the `QueryContext` layer — multiple `QueryResultImpl` instances sharing the same canonical query key now reuse one underlying `GraphQueryContext`/`QueueQueryContext` instead of spawning N separate streams
- Adds `canonicalQueryKey()` which strips cosmetic fields (`debugLabel`) and serialises with `json-stable-stringify` to produce a stable, property-order-independent key
- Wires the coalescer into `HypergraphImpl._query()` and `QueueImpl._query()`, keeping `QueryResultImpl` and the `QueryContext` interface fully unchanged
- Exposes `QueryContextCoalescer.diagnostics` (array of `CoalescerDiagnostic`) for per-key observability without global state

### Architecture

```
QueryResultImpl          ← unchanged public API
   └── CoalescedHandle   ← implements QueryContext; new per-subscriber handle
          └── SharedEntry (1 per canonical key)
                 └── GraphQueryContext / QueueQueryContext  ← single instance shared
```

Key behaviours:
- **Refcount + grace period**: underlying context starts when first handle starts, stops 200 ms after the last handle stops (absorbs React StrictMode double-mount)
- **In-flight `run()` deduplication**: concurrent one-shot calls sharing a key join the same promise; per-caller timeout applied independently; caller ctx cancellation does not abort the shared request; maximum requested timeout tracked across joined callers
- **Terminal dispose guard**: `getOrCreate()` throws after `dispose()`; `update()` and `#incrementRef()` silently return if disposed
- **Read-your-writes preserved**: no caching of resolved results; sequential `run()` calls always dispatch a fresh underlying `run()`

### Changes

- `packages/core/echo/echo-db/src/query/canonical-query-key.ts` — pure function for query key canonicalization
- `packages/core/echo/echo-db/src/query/query-context-coalescer.ts` — `QueryContextCoalescer` + `CoalescedHandle`
- `packages/core/echo/echo-db/src/hypergraph.ts` — wired into `_query()`, exposes `dispose()`
- `packages/core/echo/echo-db/src/queue/queue.ts` — wired into `_query()`
- `packages/core/echo/echo-db/src/query/canonical-query-key.test.ts` — 8 unit tests for key canonicalization
- `packages/core/echo/echo-db/src/query/query.test.ts` — 3 integration tests for coalescing behaviour

## Test plan

- [x] `moon run echo-db:test` — 564 tests pass, 0 failures
- [x] `moon run echo-db:lint` — passes
- [x] `moon run echo-db:build` — no type errors
- [x] `moon run client-services:test` — 98 tests pass (14 skipped in CI)

https://claude.ai/code/session_016k2iLAxu4BHpmKWoVjzXeu

### CODE REVIEW

File: packages/core/echo/echo-db/src/hypergraph.ts [hunk 1]
```diff
@@ -35,6 +44,17 @@ export class HypergraphImpl implements Hypergraph.Hypergraph {
...
+
+  constructor() {
+    this._coalescer = new QueryContextCoalescer(() => this._createLiveObjectQueryContext());
+    registerCoalescer(this._coalescer);
+  }
+
+  dispose(): void {
+    unregisterCoalescer(this._coalescer);
+    this._coalescer.dispose();
+  }
```
[the remark this answers is not in the document]
Comment (dmaretskyi) [reply packages/core/echo/echo-db/src/hypergraph.ts:61]: Fixed in the latest commit — `EchoClient._close()` now calls `this._graph.dispose()` at the end of its shutdown sequence, which triggers `HypergraphImpl.dispose()` → `unregisterCoalescer` + `_coalescer.dispose()`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/core/echo/echo-db/src/query/query-context-coalescer.ts [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+    if (key !== entry.key) {
+      return entry.underlying.run(ctx, ast, opts);
+    }
+
+    if (!entry.inFlight) {
+      entry.lifetimeRuns++;
+      // Use a fresh context so per-caller ctx cancellations don't abort the shared request.
+      entry.inFlight = entry.underlying.run(Context.default(), ast, { timeout: 60_000 }).finally(() => {
+        entry.inFlight = undefined;
+      });
```
[the remark this answers is not in the document]
Comment (dmaretskyi) [reply packages/core/echo/echo-db/src/query/query-context-coalescer.ts]: Fixed — the underlying run now uses `Math.max(timeout, 60_000)` so a caller with a timeout longer than 60s won't be prematurely rejected by the shared request.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/core/echo/echo-db/src/query/query-context-coalescer.test.ts [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+import { Filter, Query } from '@dxos/echo';
+import { type QueryAST } from '@dxos/echo-protocol';
+
+import { type QueryContext } from './query-context';
+import { QueryContextCoalescer } from './query-context-coalescer';
+
+/**
+ * Minimal mock implementing QueryContext.
+ */
+class MockQueryContext implements QueryContext {
```
Comment (dmaretskyi) [inline packages/core/echo/echo-db/src/query/query-context-coalescer.test.ts]: do not do mock tests, instead expand the real query.test.ts to cover coalesing edge-cases

File: packages/core/echo/echo-db/src/query/query-context-coalescer.ts [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+export const registerCoalescer = (coalescer: QueryContextCoalescer): void => {
+  COALESCERS.add(coalescer);
+};
+
+/** Unregister a coalescer instance from the global diagnostic. */
+export const unregisterCoalescer = (coalescer: QueryContextCoalescer): void => {
+  COALESCERS.delete(coalescer);
+};
+
+trace.diagnostic({
```
Comment (dmaretskyi) [inline packages/core/echo/echo-db/src/query/query-context-coalescer.ts]: those global diagnostics are depreacted at this point so remove it

File: packages/core/echo/echo-db/src/query/query-context-coalescer.ts [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+    if (key !== entry.key) {
+      return entry.underlying.run(ctx, ast, opts);
+    }
+
+    if (!entry.inFlight) {
+      entry.lifetimeRuns++;
+      // Use a fresh context so per-caller ctx cancellations don't abort the shared request.
+      entry.inFlight = entry.underlying
+        .run(Context.default(), ast, { timeout: Math.max(timeout, 60_000) })
+        .finally(() => {
```
[the remark this answers is not in the document]
Comment (dmaretskyi) [reply packages/core/echo/echo-db/src/query/query-context-coalescer.ts:144]: Fixed — added `entry.maxRequestedTimeout` (initialized to 0, reset after each in-flight settles). Before dispatch, it's updated with the current caller's timeout, and the underlying runs with `Math.max(entry.maxRequestedTimeout, 5 * 60_000)` — both the per-entry tracking and the 5-minute floor are now in place.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/core/echo/echo-db/src/query/query-context-coalescer.ts [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+      deleteEntry: (key) => this.#deleteEntry(key),
+    };
+  }
+
+  /** Returns a handle that shares an underlying context with any other handle for the same canonical key. */
+  getOrCreate(ast: QueryAST.Query): QueryContext<T, O> {
+    const handle = new CoalescedHandle<T, O>(this.#internals);
+    handle.update(ast);
+    return handle;
+  }
```
[the remark this answers is not in the document]
Comment (dmaretskyi) [reply packages/core/echo/echo-db/src/query/query-context-coalescer.ts:247]: Implemented the terminal guard: `getOrCreate()` and `#getOrCreateEntry()` throw when disposed; `CoalescedHandle.update()` and `#incrementRef()` check `isDisposed()` (via `CoalescerInternals`) to silently skip; `dispose()` now explicitly clears `stopTimer`, `isRunning`, `refcount`, and `inFlight` on each entry before clearing the map.\n\nNote: `EchoClient._close()` no longer calls `this._graph.dispose()` — `EchoClient` reuses the same `HypergraphImpl` across open/close cycles (it's a `readonly` field), so calling terminal `dispose()` on close broke client restart. The global `COALESCERS` set was already removed in a previous commit, so there's no longer a static reference leak to worry about.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by dmaretskyi** [comment]
Closing for now. While the implementation is correct and all tests pass, we don't currently have evidence that duplicate `QueryContext` instances are a meaningful source of overhead in practice. The coalescer adds non-trivial complexity (grace timers, refcounting, stale-context handling across lifecycle cycles) that we had to debug and fix during CI. Without profiling data or a measured regression to point to, the complexity cost outweighs the theoretical benefit. Can be reopened if we gather concrete perf evidence showing duplicated query contexts are a bottleneck.

---
_Generated by [Claude Code](https://claude.ai/code)_
